### DISCUSSION TITLE
Quarkus CXF Soap Xml Schema Validation

### DISCUSSION THREAD

**Discussion by cj19** [quarkiverse/quarkus-cxf#1060] [Q&A] [answered]
Hi!

In my project I try to turn on the schema validaton for a web service, but seems like it does not work.

- Imported the latest cxf dependency:
`implementation 'io.quarkiverse.cxf:quarkus-cxf:2.4.1`
- Even tried to set the schema in the app.properties:
```
schema.validation.enabled=true
quarkus.cxf.schema-validation-location
```
- Also tried to put the @SchemaValidation annotation on the web service class.

What do I do wrong?

**Comment by ppalaga** [comment]
Sounds like you use Gradle? 
We do not test with Gradle, so it is quite possible that it does not work at all. 
A minimal reproducer project would be nice.

**Reply by cj19** [reply]
Yes, I do. Unfortunately I can not provide you reproducer project.

**Reply by ppalaga** [reply]
To exclude all basic dependency management misconfigurations, please go to code.quarkus.io, select the desired extensions and let it generate a gradle project for you. Then compare the generated `build.gradle` with yours.

> ```
> schema.validation.enabled=true
> quarkus.cxf.schema-validation-location
> ```

Are you sure this is how you wanted to type it? I am not familiar with the named properties. Could you please point me to a resource where they are documented?

**Reply by cj19** [reply]
I pushed an [example-project](https://github.com/cj19/quarkus-schema-example), you can test it out with gradle. If you have any question, then please do not hesitate to ask.
`schema.validation.enabled=true
quarkus.cxf.schema-validation-location`
These config names that I found on internet, at the end I tried anything to make it work. :)

**Reply by ppalaga** [reply]
Thanks for the reproducer! It seems to be missing some dependencies

```
Starting a Gradle Daemon (subsequent builds will be faster)

> Task :compileJava FAILED
/home/ppalaga/orgs/cxf/quarkus-schema-example/src/main/java/service/AckService.java:5: error: package com.tsystems.etccm.iso12855.ack does not exist
import com.tsystems.etccm.iso12855.ack.AckServicePT;
                                      ^
/home/ppalaga/orgs/cxf/quarkus-schema-example/src/main/java/service/AckService.java:6: error: package com.tsystems.etccm.iso12855.ack does not exist
import com.tsystems.etccm.iso12855.ack.InfoExchangeAck;
                                      ^
/home/ppalaga/orgs/cxf/quarkus-schema-example/src/main/java/service/AckService.java:13: error: cannot find symbol
public class AckService implements AckServicePT{
                                   ^
  symbol: class AckServicePT
/home/ppalaga/orgs/cxf/quarkus-schema-example/src/main/java/service/AckService.java:16: error: cannot find symbol
        public String sendAck(InfoExchangeAck request) {
                              ^
  symbol:   class InfoExchangeAck
  location: class AckService
/home/ppalaga/orgs/cxf/quarkus-schema-example/src/main/java/service/AckService.java:15: error: method does not override or implement a method from a supertype
        @Override
        ^
5 errors

FAILURE: Build failed with an exception.
```

How should CXF know against which schema(s) it should validate the messages? There is perhaps something configured on `com.tsystems.etccm.iso12855.ack.AckServicePT` that I do not see?

**Reply by cj19** [reply]
you have to assemble it and make the quarkus framework generate the interfaces for you.

**Reply by ppalaga** [reply]
> you have to assemble it and make the quarkus framework generate the interfaces for you.

Would that be anything else than `./gradlew build`? Sorry, I am really useless when it comes to gradle.

**Reply by cj19** [reply]
yes, there is a gradle quarkus plugin by default, that let's you call the assemble command which generates the needed interfaces defined by xsdl/xml.

**Reply by cj19** [reply]
oh i forgot to push the gradle wrapper. @ppalaga

**Reply by cj19** [reply]
Any updates @ppalaga , or should I drop the topic?

**Reply by ppalaga** [reply]
Sorry, I was out of office.

Could you please list shell commands needed to reproduce the issue and also explain what you expect and what you see happening instead?

**Reply by cj19** [reply]
No problem.

Can we arrange a meeting to resolve this problem? It would be easier and faster.

**Reply by ppalaga** [reply]
Sorry, I'd prefer having this in written form, so that it is searchable for others in the future.

Could you please list shell commands needed to reproduce the issue and also explain what you expect and what you see happening instead?

**Reply by cj19** [reply]
./gradlew build or assemble will do the trick.

**Reply by ppalaga** [reply]
> ./gradlew build or assemble will do the trick.

Thanks. Could you please sketch what do you expect to happen and what actually happens instead?

**Reply by cj19** [reply]
As you can see there is the SchemaValidation annotation on the AckService class, that means the schema validation should be turned on. If you run the test it should fail, because it sends an xml where the apduAckCode has invalid value, but the test case passes successfully.

**Reply by ppalaga** [reply]
It looks like there is a WSDL file missing in `src/main/resources/wsdl` in the reproducer?

**Reply by cj19** [reply]
Sorry, I forgot to push it. :)

**Reply by ppalaga** [reply]
Thanks I can reproduce it now.

**Reply by ppalaga** [reply, marked as answer]
Here is the PR that fixes the reproducer: https://github.com/cj19/quarkus-schema-example/pull/1

The main missing bit was 

```
quarkus.cxf.endpoint."/ackService".wsdl=wsdl/AckService.wsdl
```

in application.properties

I need to check whether the same solution can also be applied to https://github.com/quarkiverse/quarkus-cxf/issues/557

**Reply by ppalaga** [reply]
Thanks again @cj19 for the reproducer and for accidentally helping to move https://github.com/quarkiverse/quarkus-cxf/issues/557 forward :)

**Reply by ppalaga** [reply]
Also nice to see that the wsdl2java works well with Gradle

**Reply by cj19** [reply]
Thank you for your help  as well. :) I am going to check the PR today.

**Reply by cj19** [reply]
I checked the project again and the schema validation works, but if i send a valid value, then I get an other error: 
WARN  [org.apa.cxf.pha.PhaseInterceptorChain] (vert.x-worker-thread-1) Interceptor for {http://tsystems.com/etccm/iso12855/ack}AckService#{http://tsystems.com/etccm/iso12855/ack}sendAck has thrown exception, unwinding now: org.apache.cxf.interceptor.Fault: The given SOAPAction EETSInfoExchangeAsync does not match an operation.

What could be the reason? @ppalaga 

Is it possible to turn off soap action validation?

**Reply by ppalaga** [reply]
Would you please add a reproducer to your repo?

**Reply by cj19** [reply]
I created a positive test case with [this request](https://pastebin.com/9LEPwJzu). The test case runs successfully, but if you run the whole service and try to send that request using SoapUI it fails with the error mentioned above. Very strange and can't find a solution.

**Reply by ppalaga** [reply]
Have you tried [enabling payload logging](https://docs.quarkiverse.io/quarkus-cxf/dev/reference/extensions/quarkus-cxf-rt-features-logging.html) on the service side and compare whether SoapUI sends the same request as your test?

**Reply by cj19** [reply]
Not yet. I copied out the request from the test case and copied it into SoapUI and formatted it. I mean that shouldn't be the problem, right? 🤔

**Reply by cj19** [reply]
After creating an interceptor in order to compare the messages I was able to realize that the tags I used in the xml was invalid. :D
Thank you for the help @ppalaga !

**Reply by ppalaga** [reply]
Happy to hear that!

**Reply by cj19** [reply]
Let me to ask one more question, is there a way to dynamically turn off and on the schema validation using config property? @ppalaga

**Reply by ppalaga** [reply]
No, but it could be implemented quite easily. You may want to file a [new issue](https://github.com/quarkiverse/quarkus-cxf/issues/new) if you see such a need.

**Reply by cj19** [reply]
Thank you for your suggestion/cooperation. [Here is my issue.](https://github.com/quarkiverse/quarkus-cxf/issues/1099) @ppalaga
